subsys/bluetooth/services/ots/ots_srv.c

```diff
@@ -0,0 +1,393 @@
+	/* Check OACP supported features against Kconfig. */
+	if (ots_init->features.oacp & (~((uint32_t) OACP_FEAT))) {
+		return -ENOTSUP;
+	}
+
+	ots->features.oacp = ots_init->features.oacp;
+	LOG_DBG("OACP features: 0x%04X", ots->features.oacp);
+
+	/* Check OLCP supported features against Kconfig. */
+	if (ots_init->features.olcp & (~((uint32_t) OLCP_FEAT))) {
+		return -ENOTSUP;
+	}
+	ots->features.olcp = ots_init->features.olcp;
+	LOG_DBG("OLCP features: 0x%04X", ots->features.olcp);
+
+	/* Register L2CAP context. */
+	err = bt_gatt_ots_srv_l2cap_register(&ots->l2cap);
+	if (err) {
+		return err;
+	}
+
+	err = bt_gatt_service_register(ots->service);
+	if (err) {
+		bt_gatt_ots_srv_l2cap_unregister(&ots->l2cap);
+
+		return err;
+	}
+
+	LOG_DBG("Initialized OTS");
+
+	return 0;
+}
+
+#if defined(CONFIG_BT_GATT_OTS_SRV_SECONDARY_SVC)
+	#define BT_GATT_OTS_SRV_SERVICE	BT_GATT_SECONDARY_SERVICE
+#else
+	#define BT_GATT_OTS_SRV_SERVICE	BT_GATT_PRIMARY_SERVICE
+#endif
+
+#define BT_GATT_OTS_SRV_ATTRS(_ots)	{				\
+	BT_GATT_OTS_SRV_SERVICE(BT_UUID_OTS),				\
+	BT_GATT_CHARACTERISTIC(BT_UUID_OTS_FEATURE,			\
+		BT_GATT_CHRC_READ, BT_GATT_PERM_READ,			\
+		ots_feature_read, NULL, &_ots),				\
+	BT_GATT_CHARACTERISTIC(BT_UUID_OTS_NAME,			\
+		BT_GATT_CHRC_READ, BT_GATT_PERM_READ,			\
+		ots_obj_name_read, NULL, &_ots),			\
+	BT_GATT_CHARACTERISTIC(BT_UUID_OTS_TYPE,			\
+		BT_GATT_CHRC_READ, BT_GATT_PERM_READ,			\
+		ots_obj_type_read, NULL, &_ots),			\
+	BT_GATT_CHARACTERISTIC(BT_UUID_OTS_SIZE,			\
+		BT_GATT_CHRC_READ, BT_GATT_PERM_READ,			\
+		ots_obj_size_read, NULL, &_ots),			\
+	BT_GATT_CHARACTERISTIC(BT_UUID_OTS_ID,				\
+		BT_GATT_CHRC_READ, BT_GATT_PERM_READ,			\
+		ots_obj_id_read, NULL, &_ots),				\
+	BT_GATT_CHARACTERISTIC(BT_UUID_OTS_PROPERTIES,			\
+		BT_GATT_CHRC_READ, BT_GATT_PERM_READ,			\
+		ots_obj_prop_read, NULL, &_ots),			\
+	BT_GATT_CHARACTERISTIC(BT_UUID_OTS_ACTION_CP,			\
+		BT_GATT_CHRC_WRITE | BT_GATT_CHRC_INDICATE,		\
+		BT_GATT_PERM_WRITE, NULL,				\
+		bt_gatt_ots_srv_oacp_write, &_ots),			\
+	BT_GATT_CCC_MANAGED(&_ots.oacp_ind.ccc,				\
+		BT_GATT_PERM_READ | BT_GATT_PERM_WRITE),		\
+	BT_GATT_CHARACTERISTIC(BT_UUID_OTS_LIST_CP,			\
+		BT_GATT_CHRC_WRITE | BT_GATT_CHRC_INDICATE,		\
+		BT_GATT_PERM_WRITE, NULL,				\
+		bt_gatt_ots_srv_olcp_write, &_ots),			\
+	BT_GATT_CCC_MANAGED(&_ots.olcp_ind.ccc,				\
+		BT_GATT_PERM_READ | BT_GATT_PERM_WRITE)			\
+}
+
+#define BT_GATT_OTS_SRV_ATTRS_DEF(n, _ots_instances)			\
+	static struct bt_gatt_attr ots_attrs_##n[] =			\
+		BT_GATT_OTS_SRV_ATTRS(_ots_instances[n]);
+
+#define BT_GATT_OTS_SRV_SVC_ITEM(n, _) BT_GATT_SERVICE(ots_attrs_##n),
+
+#define BT_GATT_OTS_SRV_INSTANCE_LIST_DEF(_instance_num)		  \
+	static struct bt_gatt_ots_srv ots_instances[_instance_num];	  \
+	UTIL_EVAL(UTIL_REPEAT(						  \
+		_instance_num, BT_GATT_OTS_SRV_ATTRS_DEF, ots_instances)) \
+	static struct bt_gatt_service ots_service_list[] = {		  \
+		UTIL_LISTIFY(_instance_num, BT_GATT_OTS_SRV_SVC_ITEM)	  \
+	}
```

Comment:
It would be great if we could generalize this to be able to define multiple instances in gatt.h (e.g. BT_GATT_SERVICE_INSTANCE), though it might be simpler to not have the objects such as bt_gatt_ots_srv and instead just pass the instance number to the callbacks so services can then resolve to its internal objects as they please.

Reply:
I concur, that would be a very welcome addition

Reply:
Sure, I can add it, but I would rather do it in subsequent PRs. The scope of this one is already quite big.

This general macro part could accept another parameter _instances that contains whatever the user wants as a context in the service callbacks (whether it is an array of indexes or actual instances like bt_gatt_ots_srv). We can discuss it further when the PR is open.

Reply:
Can it be added later without changing the API? Even though it is marked as experimental it is best to avoid, since it might propagate to downstream repositories that might use this API.

Reply:
It would be a new API/Macro, so we should not affect BT_GATT_SERVICE or relates APIs.

Reply:
There will be no API breaking either for OTS or in gatt.h. This instance definition macro is declared and used in the ots_srv.c file so the API user cannot access it.